cpp/src/arrow/csv/writer_test.cc

```diff
@@ -69,25 +70,49 @@ std::vector<WriterTestParams> GenerateTestCases() {
                              { "a": 124, "b\"": "a\"\"b\"" },
                              { "d": 0 },
                              { "e": 86400000 },
-                             { "f": 1078016523 }])";
-  std::string expected_without_header = std::string("1,,-1,,,") + "\n" +       // line 1
-                                        R"(1,"abc""efg",2324,,,)" + "\n" +     // line 2
-                                        R"(,"abcd",5467,,,)" + "\n" +          // line 3
-                                        R"(,,,,,)" + "\n" +                    // line 4
-                                        R"(546,"",517,,,)" + "\n" +            // line 5
-                                        R"(124,"a""""b""",,,,)" + "\n" +       // line 6
-                                        R"(,,,1970-01-01,,)" + "\n" +          // line 7
-                                        R"(,,,,1970-01-02,)" + "\n" +          // line 8
-                                        R"(,,,,,2004-02-29 01:02:03)" + "\n";  // line 9
+                             { "f": 1078016523 },
+                             { "b\"": "NA" }])";
+  std::string expected_without_header = std::string("1,,-1,,,") + "\n" +        // line 1
+                                        R"(1,"abc""efg",2324,,,)" + "\n" +      // line 2
+                                        R"(,"abcd",5467,,,)" + "\n" +           // line 3
+                                        R"(,,,,,)" + "\n" +                     // line 4
+                                        R"(546,"",517,,,)" + "\n" +             // line 5
+                                        R"(124,"a""""b""",,,,)" + "\n" +        // line 6
+                                        R"(,,,1970-01-01,,)" + "\n" +           // line 7
+                                        R"(,,,,1970-01-02,)" + "\n" +           // line 8
+                                        R"(,,,,,2004-02-29 01:02:03)" + "\n" +  // line 9
+                                        R"(,"NA",,,,)" + "\n";                  // line 10
+
   std::string expected_header = std::string(R"("a","b""","c ","d","e","f")") + "\n";
 
+  auto schema_custom_na = schema({field("g", uint64()), field("h", utf8())});
+
+  auto populated_batch_custom_na = R"([{"g": 42, "h": "NA"},
+                                        {}])";
+
+  std::string expected_custom_na = std::string(R"(42,"NA")") + "\n" +  // line 1
+                                   R"(NA,NA)" + "\n";                  // line 2
+
+  std::string expected_custom_quoted_na = std::string(R"(42,"NA")") + "\n" +  // line 1
+                                          R"(""NA"",""NA"")" + "\n";          // line 2
```

Comment:
Hmm, this doesn't look exactly right. I think it should be `"""NA"""`, not `""NA""`.

Reply:
Though it seems a bug in the escaping logic, not in this PR.

Reply:
Actually, no, the problem really seems to be in this PR, since otherwise the CSV writer works fine:
```python
>>> tab = pa.table({'a': [0,1,2], 'b': ['foo', '"bar"', '""baz""']})
>>> bio = io.BytesIO()
>>> csv.write_csv(tab, bio)
>>> bio.getvalue()
b'"a","b"\n0,"foo"\n1,"""bar"""\n2,"""""baz"""""\n'
>>> bio.seek(0)
0
>>> tt = csv.read_csv(bio)
>>> tt
pyarrow.Table
a: int64
b: string
----
a: [[0,1,2]]
b: [["foo",""bar"","""baz"""]]
>>> tab == tt
True
```

Reply:
Right. What I mean is that the result here seems wrong (in `expected_custom_quoted_na`).

Reply:
Here's with the `csv` module from the Python standard library:
```python
>>> import csv
>>> lines = ['NA,"NA",""NA"","""NA"""']
>>> csv.reader(lines)
<_csv.reader at 0x7fc19c548f90>
>>> list(csv.reader(lines))
[['NA', 'NA', 'NA""', '"NA"']]
```

As you see, the right quoting-escaping for `"NA"` shuold be `"""NA"""` (three quotes).

Reply:
So the point is that if the value that represents null is `"NA"`, if we would render it as `"""NA"""`, this would imply this is a string (because of the outer quotes) and the string value is obtained after unescaping the inner quotes, which would be would be `"NA"` (:string). But is not a string, it's `"NA"` (:null).